### ISSUE TITLE
Here geocoder authentication issues

### ISSUE THREAD

**Issue by kenseii** [geopy/geopy#381]
Maybe i am mistaken but the Here API wrapper inside geopy is outdated, it seems it cant support the `api_key` argument which is the current way of authenticating with the here api alongside the app_id.

for example this fail as the current class uses app_id and app_code:
```
from geopy.geocoders import Here as GeopyHere
client = GeopyHere(api_key="YOUR_API_KEY", app_id="YOUR_APP_ID")
client.geocode("Shimogyo")
```

Basically all new apps don't receive an `app_code` so they cant work with geopy's wrapper.
is it planned in a future release or a PR is welcome? 
[Here docs about the app_code stopping:](https://developer.here.com/documentation/authentication/dev_guide/index.html)

**Comment by enriqueav** [comment]
Just to be clear. The API still supports the old type of authentication, so if you already have created an APP with `app_id` and `app_code`, it still works.

So it seems for a period of time geopy should support
1. Old-style `app_id` and `app_code`, currently supported

**Comment by KostyaEsmukov** [comment]
> is it planned in a future release or a PR is welcome?

No. PR is welcome.

> 3. OAuth 2.0 (JSON Web Tokens). Not sure if it's supported for other geocoders

Currently there're no geocoders using oauth in geopy, but there was one before: 5fe188407118b51a0a9a11e5606b9993bbc2eba3

**Comment by deeplook** [comment]
I was just about to raise an issue myself. I can help with this as I've implemented the original plugin as well. I would add only the new APIKEY mechanism available for new HERE accounts, and leave OAuth2 for later or somebody else.

In any case it would be a breaking change to the internal API which now has two positional parameters `app_code` and `app_id`. Since these will continue to work for old accounts for an unspecified period of time (see [details](https://developer.here.com/blog/announcing-two-new-authentication-types) there is a need to support both for now. Maybe @KostyaEsmukov can suggest a clean way how this would work best inside geopy?

**Comment by KostyaEsmukov** [comment]
> Since these will continue to work for old accounts for an unspecified period of time (see [details](https://developer.here.com/blog/announcing-two-new-authentication-types)) there is a need to support both for now (`app_code` and `app_id` or only `appkey` alone). Maybe @KostyaEsmukov can suggest a clean way how this would work best inside geopy? 

Making `app_code` and `app_id` optional (with default value `None`) and adding a new keyword argument `appkey=None` should be fine. This way the old authentication method would still work in a backwards-compatible way, and the new one could be used with kwarg as `Here(appkey="...")`.

GoogleV3 implements a similar approach with a subsequent validation of passed credentials in `__init__`: https://github.com/geopy/geopy/blob/0f0ac3e030f13c804a7af7c0aa238f241bd4f589/geopy/geocoders/googlev3.py#L101-L104

@deeplook I'm going to cut a new release this week. Would you like me to wait for your PR with the fix so it could get to the upcoming release?

**Comment by deeplook** [comment]
@KostyaEsmukov If you can wait for a few days I'd appreciate. I've implemented the fix in a different context already. It's just a matter of repeating this for geopy.

**Comment by KostyaEsmukov** [comment]
@deeplook No problem at all, thank you for prompt reply!

**Comment by trevorphillips** [comment]
is there a fix for this yet?

**Comment by deeplook** [comment]
@trevorphillips On it...

**Comment by deeplook** [comment]
@trevorphillips @KostyaEsmukov See https://github.com/geopy/geopy/pull/388.

**Comment by KostyaEsmukov** [comment]
Fixed by b81969c...29b42e5

**Comment by KostyaEsmukov** [comment]
Released in 1.21.0.
